src/storage/secondary/RedisStorage.cpp

```diff
@@ -0,0 +1,353 @@
+    }
+  } else {
+    LOG("Redis invalid url: {}", m_url);
+    m_invalid = true;
+    return false;
+  }
+
+  if (!m_context || m_context->err) {
+    LOG("Redis connect {} err: {}", m_url, m_context->errstr);
+    m_invalid = true;
+    return false;
+  } else {
+    if (m_context->connection_type == REDIS_CONN_TCP) {
+      LOG(
+        "Redis connect tcp {}:{} OK", m_context->tcp.host, m_context->tcp.port);
+    }
+    if (m_context->connection_type == REDIS_CONN_UNIX) {
+      LOG("Redis connect unix {} OK", m_context->unix_sock.path);
+    }
+    m_connected = true;
+
+    if (m_operation_timeout) {
+      LOG("Redis timeout {}", timeval_to_string(*m_operation_timeout));
+      if (redisSetTimeout(m_context, *m_operation_timeout) != REDIS_OK) {
+        LOG_RAW("Failed to set timeout");
+      }
+    }
+
+    if (m_password) {
+      std::string username = m_username ? *m_username : "default";
+      LOG("Redis AUTH {} {}", username, "*****"); // don't log m_password !!!
+      redisReply* reply;
+      if (m_username) {
+        reply = static_cast<redisReply*>(redisCommand(
+          m_context, "AUTH %s %s", m_username->c_str(), m_password->c_str()));
+      } else {
+        reply = static_cast<redisReply*>(
+          redisCommand(m_context, "AUTH %s", m_password->c_str()));
+      }
+      if (!reply) {
+        LOG("Failed to auth {} in redis", username);
+        m_invalid = true;
+      } else if (reply->type == REDIS_REPLY_ERROR) {
+        LOG("Failed to auth {} in redis: {}", username, reply->str);
+        m_invalid = true;
+      }
+      freeReplyObject(reply);
+      if (m_invalid) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+}
+
+void
+RedisStorage::disconnect()
+{
+  if (m_connected) {
+    // Note: only the async API actually disconnects from the server
+    //       the connection is eventually cleaned up in redisFree()
+    LOG_RAW("Redis disconnect");
+    m_connected = false;
+  }
+}
+
+nonstd::expected<nonstd::optional<std::string>, SecondaryStorage::Error>
+RedisStorage::get(const Digest& digest)
+{
+  if (!connect()) {
+    return nonstd::make_unexpected(Error::error);
+  }
+  const std::string key = get_key_string(digest);
+  LOG("Redis GET {}", key);
+  redisReply* reply =
+    static_cast<redisReply*>(redisCommand(m_context, "GET %s", key.c_str()));
+  bool found = false;
+  bool missing;
+  std::string value;
+  if (!reply) {
+    LOG("Failed to get {} from redis", key);
+  } else if (reply->type == REDIS_REPLY_ERROR) {
+    LOG("Failed to get {} from redis: {}", key, reply->str);
+  } else if (reply->type == REDIS_REPLY_STRING) {
+    value = std::string(reply->str, reply->len);
+    found = true;
+  } else if (reply->type == REDIS_REPLY_NIL) {
+    missing = true;
+  }
+  freeReplyObject(reply);
+  if (found) {
+    return value;
+  } else if (missing) {
+    return nonstd::nullopt;
+  } else {
+    return nonstd::make_unexpected(Error::error);
```

Comment:
Is it possible to distinguish timeout from error in `redisReply`? If so, return `Error::timeout` for timeout.

Reply:
It should be possible, when using hiredis 1.0 and above. But not when using a `bool` for success

Reply:
> But not when using a bool for success

You'd return `nonstd::make_unexpected(Error::timeout)` in that case, not a bool.

Reply:
Right, mostly internal refactoring